dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/predictor/PredictionFormatterTest.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2004-2021, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.predictor;
+
+import static org.hisp.dhis.analytics.AggregationType.NONE;
+import static org.hisp.dhis.common.ValueType.BOOLEAN;
+import static org.hisp.dhis.common.ValueType.DATE;
+import static org.hisp.dhis.common.ValueType.FILE_RESOURCE;
+import static org.hisp.dhis.common.ValueType.INTEGER;
+import static org.hisp.dhis.common.ValueType.NUMBER;
+import static org.hisp.dhis.common.ValueType.TEXT;
+import static org.hisp.dhis.predictor.PredictionFormatter.formatPrediction;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+import org.hisp.dhis.DhisConvenienceTest;
+import org.hisp.dhis.dataelement.DataElement;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * @author Jim Grace
+ */
+public class PredictionFormatterTest
+    extends DhisConvenienceTest
+{
+    private DataElement deNumber;
+
+    private DataElement deInteger;
+
+    private DataElement deText;
+
+    private DataElement deBoolean;
+
+    private DataElement deDate;
+
+    private DataElement deFileResource;
+
+    @Before
+    public void initTest()
+    {
```

Comment:
I tend to avoid init() methods unless you need some common setup/behaviour that is needed by all unit test methods. Otherwise, I would suggest moving the mock/stub date to the unit test they actually belong to.

This will make the tests more readable (we will know exactly the fixtures and assertions for each unit test), and will avoid setting up data that is not necessary for all unit tests.
I'm afraid we can end up on this kind of situation: https://github.com/dhis2/dhis2-core/blob/0888f55a0f2272e80a7044dc959f93679636f74f/dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/ValidationServiceTest.java#L237
We have no idea what data is used by which unit test and which ones are being shared. Maintaining these kinds of tests can become painful.

Is easier to read and maintain something more structured and concise where we can see the full scope of test in one place. Something like:

Unit test A
// Fixtures
// The actual test
// The assertions
...

Unit test X
// Fixtures
// The actual test
// The assertions

This is just a suggestion. I leave the decision to you.

Reply:
@maikelarabori Point well taken, especially since there were no shared fixtures between these tests. I've changed it as you suggested. Thanks!